modules/dnn/src/layers/einsum_layer.cpp

```diff
@@ -202,12 +191,198 @@ bool IsTransposeRequired(size_t input_rank, const std::vector<size_t>& permutati
   return transpose_required;
 }
 
-Mat Diagonal(
-    const cv::Mat& input,
-    int subscriptIndicesToInputIndex,
-    int dimIndexInIreprocessedInput)
+
+bool IsTransposeRequiredForDiagonal(int dim1, int dim2, int rank) {
+    // If the input is 2D, we don't need a transpose
+    if (rank == 2)
+        return false;
+
+    // If the two dims are the innermost dims, no transpose is required
+    if ((dim1 == rank - 1 && dim2 == rank - 2) ||
+        (dim1 == rank - 2 && dim2 == rank - 1))
+        return false;
+
+    // Transpose is required
+    return true;
+}
+
+template <typename T>
+Mat DiagonalDataAssignment(Mat input) {
+
+    int rank = input.dims;
+    CV_Assert(rank >= 2);
+    CV_Assert(input.size[rank - 1] == input.size[rank - 2]);
+    MatShape original_dims = shape(input);
+
+    if (rank > 3){
+        //reshape to 3D mat
+        int collapsed_size = 1;
+        for (int i = 0; i < rank - 2; ++i) {
+            collapsed_size *= input.size[i];
+        }
+        std::vector<int> reshaped_dims = {collapsed_size, input.size[rank - 2], input.size[rank - 1]};
+        input = input.reshape(1, reshaped_dims);
+    }
+
+    // Compute total number of higher-dimensional slices
+    int total_slices = input.size[0];
+
+    // Create the output matrix (keeping the original dimensions but changing the size of the last dimension to 1)
+    std::vector<int> output_dims(input.dims);
+    for (int i = 0; i < input.dims - 1; ++i) {
+        output_dims[i] = input.size[i];
+    }
+    output_dims.back() = 1;
+    Mat output = Mat(output_dims, input.type());
+
+    int inner_stride = input.size[input.dims - 1];
+    for (int slice = 0; slice < total_slices; ++slice) {
+        for (int j = 0; j < inner_stride; ++j) {
+            // Extract diagonal elements
+            output.at<T>(slice, j, 0) = input.at<T>(slice, j, j);
+        }
+    }
+
+    // Reshape output back to original shape
+    original_dims[rank - 1] = 1;  // Set the last dimension to 1, as we have extracted the diagonal
+    output = output.reshape(1, original_dims);
+    return output;
+}
+
+// Extract the diagonal elements from the last two dimensions of the tensor.
+// For instance, given an input_shape of [1, 2, 3, 3]:
+
+// The flexibility in this implementation allows one to choose which of the two
+// last dimensions retains its value, determined by the `preserve_innermost_dim_val` parameter.
+
+// When preserve_innermost_dim_val == true:
+//       The resulting shape is [1, 2, 1, 3], indicating the diagonal has 3 elements,
+//       and it keeps the dimension value of the innermost dimension.
+
+// When preserve_innermost_dim_val == false:
+//       The resulting shape is [1, 2, 3, 1], indicating the diagonal also has 3 elements,
+//       but it retains the dimension value of the penultimate dimension.
+Mat DiagonalInnermostDims(const Mat& input, bool preserve_innermost_dim_val) {
```

Comment:
All these comments are for `DiagonalInnermostDims` right? I propose to use `/**/` for these comments to avoid misleading.

Reply:
done